File: 88-merge-sorted-array/88-merge-sorted-array.py
class Solution(object):
    def merge(self, nums1, m, nums2, n):
        """
        :type nums1: List[int]
        :type m: int
        :type nums2: List[int]
        :type n: int
        :rtype: None Do not return anything, modify nums1 in-place instead.
        """
        i = m-1
        j = n-1
        indexreq = m+n-1
        
        while i>=0 and j>=0:
            if nums1[i] > nums2[j]:
                nums1[indexreq] = nums1[i]
                i -= 1
                indexreq -= 1
            elif nums2[j] >= nums1[i]:
                nums1[indexreq] = nums2[j]
                j -= 1
                indexreq -= 1
            else:
                nums1[indexreq] = nums2[j]
                nums1[indexreq-1] = nums1[j]
                i -= 1
                j -= 1
                indexreq -= 2
        
        while j>=0:
            nums1[indexreq] = nums2[j]
            j -= 1
            indexreq -= 1
        
        # Brute Force Approach
        # if m<1:
        #     nums1[:] = nums2[:n]
        # elif n<1:
        #     nums1[:] = nums1[:m]
        # else:
        #     nums1[:] = sorted(nums1[:m]+nums2[:n])

File: 88-merge-sorted-array/88-merge-sorted-array.java
class Solution {
    public void merge(int[] nums1, int m, int[] nums2, int n) {
        int i = m-1;
        int j = n-1;
        int indreq = m+n-1;
        
        while(i>=0 && j>=0){
            if(nums1[i] > nums2[j]){
                nums1[indreq] = nums1[i];
                i--;
                indreq--;
            }
            else if(nums2[j] > nums1[i]){
                nums1[indreq] = nums2[j];
                j--;
                indreq--;
            }
            else{
                nums1[indreq] = nums1[i];
                nums1[indreq-1] = nums2[j];
                i--; 
                j--;
                indreq -= 2;
            }
        }
        
        while(j>=0){
            nums1[indreq] = nums2[j];
            j--; 
            indreq--;
        }
    }
}